### PR TITLE
Added new events for chat completions

### PR DESCRIPTION
**Pull request by nhoffmann-newrelic**

### CODE REVIEW

File: src/nr_openai_observability/build_events.py [hunk 1]
```diff
@@ -33,6 +31,27 @@ def build_events(response, request, response_headers):
...
+        else None,
+        "ratelimit_reset_tokens": ratelimit_reset_tokens,
+        "ratelimit_reset_requests": ratelimit_reset_requests,
+        "ratelimit_remaining_tokens": int(ratelimit_remaining_tokens)
+        if ratelimit_remaining_tokens.isdigit()
+        else None,
+        "ratelimit_remaining_requests": int(ratelimit_remaining_requests)
+        if ratelimit_remaining_requests.isdigit()
+        else None,
+    }
```
Comment (ykriger-newrelic) [inline src/nr_openai_observability/build_events.py:49]: maybe like that?
```
def to_int_if_digit(s):
    return int(s) if s.isdigit() else None

return {
    "ratelimit_limit_requests": to_int_if_digit(ratelimit_limit_requests),
    "ratelimit_limit_tokens": to_int_if_digit(ratelimit_limit_tokens),
    "ratelimit_reset_tokens": ratelimit_reset_tokens,
    "ratelimit_reset_requests": ratelimit_reset_requests,
    "ratelimit_remaining_tokens": to_int_if_digit(ratelimit_remaining_tokens),
    "ratelimit_remaining_requests": to_int_if_digit(ratelimit_remaining_requests),
}
```

File: src/nr_openai_observability/build_events.py [hunk 2]
```diff
@@ -49,24 +68,12 @@ def build_events(response, request, response_headers):
         "api_type": response.api_type,
         "vendor": "openAI",
```
Comment (ykriger-newrelic) [inline src/nr_openai_observability/build_events.py:69]: Should be `OpenAI` what do we have in the Node SDK?

File: src/nr_openai_observability/build_events.py [hunk 3]
```diff
@@ -76,7 +83,7 @@ def build_events(response, request, response_headers):
     return {"messages": messages, "completion": completion}
 
 
-def build_error_events(request, error):
+def build_completion_error_events(request, error):
     completion_id = str(uuid.uuid4())
```
Comment (ykriger-newrelic) [inline src/nr_openai_observability/build_events.py:87]: We should put it under a function and use it across the code

File: src/nr_openai_observability/build_events.py [hunk 4]
```diff
@@ -103,3 +110,45 @@ def build_error_events(request, error):
...
+        "id": embedding_id,
+        "api_key_last_four_digits": f"sk-{response.api_key[-4:]}",
+        "timestamp": datetime.now(),
+        "response_time": response.response_ms,
+        "request.model": request.get("model"),
+        "response.model": response.model,
+        "usage.total_tokens": response.usage.total_tokens,
+        "usage.prompt_tokens": response.usage.prompt_tokens,
+        "api_type": response.api_type,
+        "vendor": "openAI",
```
Comment (ykriger-newrelic) [inline src/nr_openai_observability/build_events.py:128]: OpenAI + having it as Enum

Comment (nhoffmann-newrelic) [reply src/nr_openai_observability/build_events.py:128]: this is the same value used in the node SDK (uncapitalised), and also having an Enum for 1 value doesn't make the most sense, we can always use. an enum if we need more values there in the future

File: src/nr_openai_observability/build_events.py (hunk 2 above)
Comment (nhoffmann-newrelic) [reply src/nr_openai_observability/build_events.py:69]: This is the value in the node SDK as I commented below

File: src/nr_openai_observability/build_events.py (hunk 1 above)
Comment (nhoffmann-newrelic) [reply src/nr_openai_observability/build_events.py:49]: will do

File: src/nr_openai_observability/build_events.py (hunk 3 above)
Comment (nhoffmann-newrelic) [reply src/nr_openai_observability/build_events.py:87]: will do

**Review by ykriger-newrelic — APPROVED** [review]

File: src/nr_openai_observability/build_events.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
+import uuid
+from datetime import datetime
+
+import openai
+
+
+def _build_messages_events(messages, completion_id, model):
+    message_id = str(uuid.uuid4())
```
Comment (RyanKadri) [inline src/nr_openai_observability/build_events.py:8]: Wanted to check my understanding here. Should we be generating one UUID for each message here vs a UUID for all messages in the "completion"?

Comment (nhoffmann-newrelic) [reply src/nr_openai_observability/build_events.py:8]: We have a completionID, which is a UUID generated for each completion.

The question of whether or not we should be generating a UUID for each message event, or can we relay on completionID+sequence is debatable, and can be changed.

**Pull request merged**